### ISSUE TITLE
Implement set, set! and copy

### ISSUE THREAD

**Issue by tlienart** [tlienart/AnalyticalEngine.jl#1]
Every leaf type of `SupervisedModel` should have a `copy` a `set` and a `set!` function. This will help with things like hyperparameter tuning where multiple copies of a base model and adjust the parameters

* [x] **copy**: take an instance and create another instance with the same parameters. There should be a way to do this without having to rewrite it explicitly for every type. 

* [x] **set!**: take an instance and mutate given parameters. Possibly "tunable" should be a function so that we can check the argumetnts are valid

something like 
```julia
function set!(glr::GLR, kwargs...)
    # go over kwargs, modify glr
    return glr
end
```

* [x] **set**: take an instance, make a copy with modified parameters. Same as `set!(copy())`

```julia
set(glr::GLR, kwargs...) = set!(copy(glr), ...)
```

**Comment by fkiraly** [comment]
Instead of re-writing this for each leaf type, could this not simply be "inherited" per dispatch from the root type, using a parameter dictionary which sits with the learner type?

I.e., the only thing changing in the descendant types would be the parameter dictionary, whereas copy and set and tune/tuneable behave the same way for all descendants? 

This would also be better from the perspective of a user defining custom learners, as they will then only have to specify fit, predict, and hyperparameter dictionary.

**Comment by tlienart** [comment]
So I think what you're suggesting is as follows:

```julia
mutable struct Model 
    hyperparams::Dict
    # ... 
end

Model({:lambda=>5.0, :penalty=>L2Penalty()})
```

etc. 

I think that could be one approach but it's a big inelegant in my opinion because to access things you have to do

```julia
print(Model.hyperparams[:lambda])
print(Model.non_hyper_param)
```

so it's not consistent from the user perspective. Also once you have an instance, it's a bit annoying to query the parameters (e.g. if you do `instance.[TAB]`). And finally you have to make sure (somehow) that the dictionary that is passed is "right". 

Maybe there's work arounds, I need to think about it a bit more

**Comment by fkiraly** [comment]
You can still have top-level variables, what I meant with "hyperparameter dictionary" is just a list of names and properties of hyperparameters (such as in mlr) rather than a dictionary of the hyperparameter values themselves. 

Making a list of hyperparameter names and properties queriable as a uniform interface point is useful for methods which in a sub-routine "copy all hyperparameters" or "tune all settable hyperparameters", especially if the hyperparameters have properties besides the value they take in an instance (e.g., lower/upper bound; queriable default; tunable yes/no).

Regarding elegance/inelegance of an approach, that always needs to be compared to alternatives, and pretty print functions are always something one can write on top of a good API design, similarly syntactic shortcuts such as Model.lambda.

**Comment by tlienart** [comment]
Ok so I don't know how to do this in Julia well @MikeInnes I think your help would be very valuable here

Some considerations;
* as far as I'm aware you don't directly have access to struct fields so it's not straightforward (IMO) to define a copy method that would "just work" will all models. 
* if we keep a dictionary that is effectively redundant of the information in the struct then it needs to contain the pointers to the values or something, I could not immediately make it such that when you modified values in the dictionary it modified values in the struct. 

Draft attempt below (because it may help explain what we're trying to do)

```julia
import Base.copy

mutable struct Foo
	hp1::Float64
	hp2::Vector{Float64}
	fp1::Bool
	fp2::Vector{Float64}
	hparams::Dict{Symbol, Ref}
end
Foo(;hp1=.0, hp2=[.0, .0], fp1=false, fp2=[1.1, 1.1]) =
	Foo(hp1, hp2, fp1, fp2, Dict(:hp1=>Ref(hp1), :hp2=>Ref(hp2)))

tuneable(f::Foo) = keys(f.hparams)

copy(f::Foo) = Foo(map(deepcopy, (f.hp1, f.hp2, f.fp1, f.fp2, f.hparams))...)

function set!(f::Foo, d::Dict)
	tuneables = tuneable(f)
	for kv in d
		# to clean up later
	 	@assert kv[1] in tuneables "given key $(kv[1]) is not a hyperparam"
                # here modify the corresponding place in f (I don't know how to do this)
	end
end
```

In any case this is quite important to fix nicely in order to

1. expose "tuneable parameters" well
2. don't have redundant structure (I don't think it's the right way to go to keep a dictionary for each type of the parameters and some pointer crap)
3. yes potentially it means that someone writing a new model has to implement `fit!`, `predict` but also a few additional function like `copy`, `set!`, `set`, `tuneable` with the usual problem in Julia that you can't guarantee that a subtype implements specific functions.

**Comment by tlienart** [comment]
Ok so this is a working implementation which may be a bit ugly.

```julia
import Base.deepcopy
import Base.setindex!

mutable struct Bar
	hparam1::Base.RefValue  # hyperparameter 1 
	hparam2::Base.RefArray  # hyperparameter 2
	fparam1::Bool # a fixed parameter
end

# function that returns symbols corresponding to hyperparameters
hparams(b::Bar) = Dict{Symbol, Ref}(
	:hparam1=>b.hparam1,
	:hparam2=>b.hparam2)

#  a constructor where you don't have to call Ref
Bar(hp1::Float64, hp2::Vector{Float64}, b::Bool) = Bar(Ref(hp1), Ref(hp2), b)

# deep copy of a model
deepcopy(b::Bar) = Bar(map(deepcopy, (b.hparam1, b.hparam2, b.fparam1))...)

# that's just to be able to mutate the elements of Base.RefArray in one shot
# needed because Base.RefArray is immutable -> need to go element by element (makes sense)
function setindex!(ra::Base.RefArray, x::AbstractArray)
	@assert length(ra.x) == length(x)
	for i in 1:length(x)
		ra.x[i] = x[i]
	end
	ra
end

# this is now fairly generic and could be applied to all models
function set!(b::Bar, d::Dict)
	hp = hparams(b) # retrieve the hyperparameters
	@assert issubset(keys(d), keys(hp)) "some keys don't match hyperparams"
	for kv in d
		setindex!(hp[kv[1]], kv[2])
	end
end

# this is a convenient simple function that creates a new object with a set based
# on a model `b` this is a use case that was discussed with fkiraly => good for 
# hyperparam tuning, stacking etc.
set(b::Bar, d::Dict) = (bn = deepcopy(b); set!(bn, d); bn)

# ====================================================================

b = Bar(1.0, [1.0, 2.0], true)

b_copy = deepcopy(b)
@show b_copy.hparam1.x
@show b_copy.hparam2.x

set!(b_copy, Dict(:hparam1=>3.0, :hparam2=>[-1.0, -1.0]))

@show b_copy.hparam1.x
@show b_copy.hparam2.x
@show b.hparam2.x

b_other = set(b, Dict(:hparam1=>5.0, :hparam2=>[0.0, 0.0]))

@show b_other.hparam2.x
@show b.hparam2.x
```

output:

```
b_copy.hparam1.x = 1.0
b_copy.hparam2.x = [1.0, 2.0]
b_copy.hparam1.x = 3.0
b_copy.hparam2.x = [-1.0, -1.0]
b.hparam2.x = [1.0, 2.0]
b_other.hparam2.x = [0.0, 0.0]
b.hparam2.x = [1.0, 2.0]
```

which is what we would want

## comments

I think that implementation will clarify what we're hoping to do, this mechanism allows to generate models out of a "blueprint" model with the hyperparameters that's changing.

So that could now be easily combined with a mechanism that generates valid dictionaries of hyperparameters drawn from some grid or other mechanism that generates sets of hyperparams (e.g. randomised grid search).

Also, I think in practice we won't need the gymnastic with `RefArray`, it will only be `RefValue` (e.g. `RefValue{Loss}` or `RefValue{Penalty}` so that funky `setindex!(ra, x)` won't be needed.

**Comment by tlienart** [comment]
I think that implementation will clarify what we're hoping to do, this mechanism allows to generate models out of a "blueprint" model with the hyperparameters that's changing.

So that could now be easily combined with a mechanism that generates valid dictionaries of hyperparameters drawn from some grid or other mechanism that generates sets of hyperparams (e.g. randomised grid search).

**Comment by tlienart** [comment]
Also, I think in practice we won't need the gymnastic with `RefArray`, it will only be `RefValue` (e.g. `RefValue{Loss}` or `RefValue{Penalty}` so that funky `setindex!(ra, x)` won't be needed.

**Comment by tlienart** [comment]
This is now implemented in b626f959852c41cda02bb2712d3201c89363f292 

It works quite well and I can see the philosophy around using Models as "MetaData containers" about what needs to be done that can be sent around easily.

One thing that I need to think about is that if we change the type of -say- Penalty, then the object fundamentally changes (if we're thinking about a `GLR{L, P}`) so this may be an argument in favour of dropping the parametric type for GLR as was advocated by Franz. 

Beyond that, it may be required to define something like a `LossRef` and `PenaltyRef` so that we can modify the value that is being pointed at. 

Maybe to clarify, the following code works

```julia
lasso = LassoRegression()
lasso3 = deepcopy(lasso)
set!(lasso3, Dict(:penalty=>2.0*L1Penalty()))
@test get(lasso3.penalty).scale == 2.0
```

while this does not

```julia
ridge_from_lasso = deepcopy(lasso)
set!(ridge_from_lasso, Dict(:penalty=>2.0*L2Penalty()))
```

error is quite clear, the `RefValue` is wrapped around a `LPPenalty{1}` and therefore does not accept a `LPPenalty{2}`. There should be an easy way around but needs a bit of thinking to make it nice.

```julia
ERROR: MethodError: Cannot `convert` an object of type AnalyticalEngine.ScaledPenalty{AnalyticalEngine.LPPenalty{2}} to an object of type AnalyticalEngine.ScaledPenalty{AnalyticalEngine.LPPenalty{1}}
This may have arisen from a call to the constructor AnalyticalEngine.ScaledPenalty{AnalyticalEngine.LPPenalty{1}}(...),
since type constructors fall back to convert methods.
```

**Comment by tlienart** [comment]
After more thinking

* removed the parametrisation for `GLR` (@fkiraly I hope this earns me a thumbs up)
* simplified a lot the way the `set`, `set!`, `deepcopy` etc works, the Pointer gig was good fun but unnecessary.
* problem above is solved since it's not a parametric type you can now write

```
lasso = LassoRegression()
ridge_from_lasso = set(lasso, penalty=L2Penalty()*2.0)
```

ee5d8f4
